flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientProvider.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.client.program.ClusterClient;
+import org.apache.flink.client.program.ClusterClientProvider;
+import org.apache.flink.client.program.rest.RestClusterClient;
+import org.apache.flink.configuration.Configuration;
+
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.apache.hadoop.yarn.api.records.ApplicationReport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.flink.yarn.YarnClusterDescriptor.setClusterEntrypointInfoToConfig;
+
+/** The implementation of ClusterClientProvider. */
+@Internal
+public class YarnClusterClientProvider implements ClusterClientProvider {
+    private static final Logger LOG = LoggerFactory.getLogger(YarnClusterClientProvider.class);
+
+    private final Object lock = new Object();
+
+    private volatile ApplicationId applicationId;
```

Comment:
I wouldn't encode whether the application report has been retrieved or not via the `ApplicationId` field. This works but an explicit state or the final `ApplicationReport` would be better. We could also think about letting the `ApplicationReportProvider` cache the `ApplicationReport`.